File: days/day11.exs
#!/usr/bin/env elixir

defmodule Day11 do
  defp evolve(stones, 0) do
    stones
  end

  defp evolve(stones, rounds) do
    state =
      stones
      |> Stream.flat_map(&split_stone/1)
      |> Enum.reduce(%{}, fn {k, v}, acc ->
        Map.update(acc, k, v, &(&1 + v))
      end)

    evolve(state, rounds - 1)
  end

  defp split_stone({0, n}) do
    [{1, n}]
  end

  defp split_stone({x, n}) do
    s = to_string(x)
    t = String.length(s)

    if rem(t, 2) == 0 do
      t = div(t, 2)

      0..1
      |> Enum.map(&String.slice(s, &1 * t, t))
      |> Enum.map(&String.to_integer/1)
      |> Enum.map(&{&1, n})
    else
      [{x * 2024, n}]
    end
  end

  def part1(data, rounds \\ 25) do
    stones =
      String.split(data)
      |> Enum.map(&String.to_integer/1)
      |> Enum.reduce(%{}, &Map.put(&2, &1, 1))

    evolve(stones, rounds)
    |> Enum.reduce(0, &(&2 + elem(&1, 1)))
  end

  def part2(data) do
    part1(data, 75)
  end
end

data = """
125 17
"""

case Day11.part1(data), do: (55312 -> true)

data = File.read!("day11.in")
ans = Day11.part1(data)
IO.puts("part1: #{ans}")
ans = Day11.part2(data)
IO.puts("part2: #{ans}")

File: days/day11.swift
#!/usr/bin/env swift
import Foundation


func part1(_ data: String, N: Int = 25) -> Int {
    var stones: [Int:Int] = data.matches(of: #/\d+/#)
        .reduce(into: [:], { acc, m in acc[Int(m.0)!] = 1 })
    for _ in 0..<N {
        var state: [Int:Int] = [:]
        for (x, n) in stones {
            if x == 0 {
                state[1, default: 0] += n
            }
            else {
                let s = String(x)
                let ix = s.indices
                if ix.count % 2 == 0 {
                    let h = s.index(s.startIndex, offsetBy: ix.count / 2)
                    let a = Int(s[..<h])!
                    let b = Int(s[h...])!
                    state[a, default: 0] += n
                    state[b, default: 0] += n
                }
                else {
                    state[x * 2024, default: 0] += n
                }
            }
        }
        stones = state
    }
    let ans = stones.reduce(0, { acc, p in acc + p.value })
    return ans
}


func part2(_ data: String) -> Int {
    part1(data, N: 75)
}


let test = """
125 17
"""
assert(part1(test) == 55312)


let data = try! String(contentsOfFile: "day11.in", encoding: .utf8)
let ans1 = part1(data)
print("part1: \(ans1)")
let ans2 = part2(data)
print("part2: \(ans2)")
